src/inabox/inabox-viewport.js

```diff
@@ -149,35 +151,91 @@ export class ViewportBindingInabox {
     /** @private @const {boolean} */
     this.useLayers_ = isExperimentOn(this.win, 'layers');
 
+    /** @private {?../../ads/inabox/position-observer.PositionObserver} */
+    this.topWindowPositionObserver_ = null;
+
     dev().fine(TAG, 'initialized inabox viewport');
   }
 
   /** @override */
   connect() {
-    this.listenForPosition_();
+    if (this.canInspectTopWindow_()) {
+      this.listenForPositionSameDomain_();
+    }
+    else {
+      this.listenForPosition_();
+    }
+  }
+
+  /**
+   * @private
+   * @param {!../layout-rect.LayoutRectDef} viewportRect
+   * @param {!../layout-rect.LayoutRectDef} targetRect
+   */
+  updateLayoutRects_(viewportRect, targetRect) {
+    const oldViewportRect = this.viewportRect_;
+    this.viewportRect_ = viewportRect;
+    this.updateBoxRect_(targetRect);
+    if (isResized(this.viewportRect_, oldViewportRect)) {
+      this.resizeObservable_.fire();
+    }
+    if (isMoved(this.viewportRect_, oldViewportRect)) {
+      this.fireScrollThrottle_();
+    }
   }
 
   /** @private */
   listenForPosition_() {
-
     this.iframeClient_.makeRequest(
         MessageType.SEND_POSITIONS, MessageType.POSITION,
         data => {
           dev().fine(TAG, 'Position changed: ', data);
-          const oldViewportRect = this.viewportRect_;
-          this.viewportRect_ = data['viewportRect'];
+          this.updateLayoutRects_(data['viewportRect'], data['targetRect']);
+        });
+  }
 
-          this.updateBoxRect_(data['targetRect']);
+  /**
+   * Returns true if top win's properties can be accessed and win is defined.
+   * This functioned is used to determine if the host window is cross-domained
+   * from the perspective of the current window.
+   * @return {boolean}
+   * @private
+   */
+  canInspectTopWindow_() {
+    // TODO: This should be changed in the future.
+    // See https://github.com/google/closure-compiler/issues/3156
+    try {
+      // win['test'] could be truthy but not true the compiler shouldn't be able
+      // to optimize this check away.
+      return !!this.win.top.location.href && (this.win.top['test'] || true);
+    } catch (unusedErr) { // eslint-disable-line no-unused-vars
+      return false;
+    }
+  }
```

Comment:
This is copied from `inabox-messaging-host.js`: can we share the code instead?

Reply:
Moved the function into `src/iframe-helper.js`, since that seems to be the most logical location.